### DISCUSSION TITLE
Re-Orient Tooltip based on mouse position

### DISCUSSION THREAD

**Discussion by arthurgailes** [visgl/deck.gl#7024] [Q&A]
Hello,

The tooltip function on the map works great, but can expand outside of the viewport when it's too far to the right or bottom. I've tried using the style prop to override the positioning, but I run into two problems:

1. The style prop doesn't override the default transform prop, and using padding or top/bottom seem to expand the width of the tooltip.
2. I don't see a way to reliably deduce the width of the tooltip from the provided object info.

Has anybody had success flipping the orientation of the tooltip relative to the mouse?

My tooltip code:
```
const flipTooltip = ({ x, y, viewport }) => {
  const { width, height } = viewport;
  const x1 = x > width * 0.75 ? x - 300 : x;
  const y1 = y > height * 0.75 ? y - 300 : y;
  return `translate(${x1}px, ${y1}px)`;
};

export function getTooltip(full) { 

  return {
    html: 'a string',
    style: { transform:  flipTooltip(full) },
  };
```

**Comment by Pessimistress** [comment]
1. We could make it so that https://github.com/visgl/deck.gl/blob/4114ace0dc9679521675bb4db6a6bacefa4fd32e/modules/core/src/lib/tooltip.ts#L85 extends `style.transform` instead of overwriting it. Care to open a PR?

2. A common technique is to create an out-of-view DOM element. You set the desired style and content on this element first and get its clientWidth and clientHeight.

**Comment by nvn-nil** [comment]
This would be quite useful. I'm planning to implement a custom context menu in my app and this would be something to solve.

> 1. We could make it so that https://github.com/visgl/deck.gl/blob/4114ace0dc9679521675bb4db6a6bacefa4fd32e/modules/core/src/lib/tooltip.ts#L85
>     extends `style.transform` instead of overwriting it. Care to open a PR?

I can open a PR for this if that's okay

> 2\. A common technique is to create an out-of-view DOM element. You set the desired style and content on this element first and get its clientWidth and clientHeight.

This is not difficult to do on my side but do you think there's merit to doing this in deck? I feel this should be the default behavior even when getTooltip returns a string.

**Reply by Pessimistress** [reply]
> I can open a PR for this if that's okay

Please do!

**Reply by nvn-nil** [reply]
Opened PR #7911

**Reply by Pessimistress** [reply]
Published v8.9.18
